cassandra/cluster.py

```diff
@@ -2596,6 +2598,18 @@ def __init__(self, session, message, query, default_timeout=None, metrics=None,
         self._errors = {}
         self._callbacks = []
         self._errbacks = []
+        self._start_timer()
+
+    def _start_timer(self):
+        if self.timeout is not None:
+            self._timer = self.session.cluster.connection_class.create_timer(self.timeout, self._on_timeout)
+
+    def _cancel_timer(self):
+        if self._timer:
+            self._timer.cancel()
+
+    def _on_timeout(self):
+        self._set_final_exception(OperationTimedOut())
```

Comment:
The ``OperationTimedOut()`` should be populated with ``_errors`` and ``_current_host``, I think.